File: CubicPath.java
package dymetric;

import java.awt.*;

public class CubicPath extends Canvas {

    protected Color ball_colour;
    protected int x_start = 50;
    protected int x_max = 200;
    protected int y_max = 10;
    protected int x_min = 500;
    protected int y_min = 310;
    protected int x = x_start;
    protected int y;
    protected double a, b, c, d; // coefficients and constant
    protected final int aWIDTH, aHEIGHT;

    public CubicPath() {
        setBackground(Color.LIGHT_GRAY); // canvas color
        ball_colour = Color.RED;
        aWIDTH = aHEIGHT = 10;

        // constants
        a = (double) (-2 * (y_max - y_min)) / Math.pow((x_max - x_min), 3);
        b = -((double) 3 / 2) * a * (x_max + x_min);
        c = 3 * a * x_max * x_min;
        d = y_max + ((double) a / 2) * (x_max - 3 * x_min) * Math.pow(x_max, 2);

        y = (int)Math.round(a * Math.pow(x, 3) + b * Math.pow(x, 2) + c * x + d);
    }

    // Feel free to double buffer if flickering occurs
    public void paint(Graphics g) {
        g.setColor(ball_colour);
        // draw a dot
        g.fillOval(x, y, aWIDTH, aHEIGHT);
    }

    public void moveCubic() {
        // condition for continuing motion
        while (x <= 750 && y >= y_max) {
            paint(this.getGraphics());
            
            x += 20;
            y = (int)Math.round(a * Math.pow(x, 3) + b * Math.pow(x, 2) + c * x + d);
            // introduce a delay betweeen renderings
            try {
                Thread.sleep(50);
            } catch (InterruptedException e) {
            }
        }
    }
}

File: CubicPath.cpp
#include "stdafx.h"
#include "CubicPath.h"
#include <math.h>


CubicPath::CubicPath(HWND hWnd, int window_width, int window_height)
{
	hWindow = hWnd; // save away window handle
	this->window_width = window_width; // save away window width
	this->window_height = window_height; // save away window width

	ball_colour = RGB(255, 0, 0);
	x_start = 50;
	x_max = 200;
	y_max = 110;
	x_min = 500;
	y_min = 410;
	x = x_start;

	// constants
	a = (double)(-2 * (y_max - y_min)) / pow((x_max - x_min), 3);
	b = -((double)3 / 2) * a * (x_max + x_min);
	c = 3 * a * x_max * x_min;
	d = y_max + ((double)a / 2) * (x_max - 3 * x_min) * pow(x_max, 2);

	y = (int)round(a * pow(x, 3) + b * pow(x, 2) + c * x + d);

	// Pen and brush for travelling ball
	ball_pen = CreatePen(PS_SOLID, 1, RGB(0, 0, 0));
	ball_brush = CreateSolidBrush(ball_colour);

	hdc = GetDC(hWindow);

	SelectObject(hdc, ball_pen); // select ball pen colour
	SelectObject(hdc, ball_brush); // select ball brush colour
}

/*
* draws the ball/circle using the apt color
*/
void CubicPath::paint() {
	// draw a dot
	Ellipse(hdc, x, y, x + aWIDTH, y + aHEIGHT);
}

/*
Repeatedly draws ball so as to simulate a continuous motion
*/
void CubicPath::moveCubic() {
	// condition for continuing motion
	while (x + aWIDTH <= window_width && y >= y_max) {
		paint();

		x += 20;
		y = (int)round(a * pow(x, 3) + b * pow(x, 2) + c * x + d);
		// introduce a delay betweeen renderings
		Sleep(50);
	}
}


CubicPath::~CubicPath()
{
	DeleteObject(ball_pen);
	DeleteObject(ball_brush);

	ReleaseDC(hWindow, hdc);
}
